Machine-Teaching-Examples/moab_state_transform.ink

```diff
@@ -0,0 +1,157 @@
+
+# What velocity do we want the ball to have when it reaches the target?
+# (This could be configurable in the sim to make it non-constant. Kept here for simplicity.)
+const TargetVelocityX = 0
+const TargetVelocityY = 0
+
+# State received from the simulator after each iteration. It includes absolute positions and velocities.
+type SimState {
+    # Ball X,Y position
+    ball_x: number<-MaxDistancePerStep - RadiusOfPlate .. RadiusOfPlate + MaxDistancePerStep>,
+    ball_y: number<-MaxDistancePerStep - RadiusOfPlate .. RadiusOfPlate + MaxDistancePerStep>,
+
+    # Ball X,Y velocity
+    ball_vel_x: number<-MaxVelocity .. MaxVelocity>,
+    ball_vel_y: number<-MaxVelocity .. MaxVelocity>,
+
+    # Target stationary X,Y position
+    target_x: number<-MaxDistancePerStep - RadiusOfPlate .. RadiusOfPlate + MaxDistancePerStep>,
+    target_y: number<-MaxDistancePerStep - RadiusOfPlate .. RadiusOfPlate + MaxDistancePerStep>,
+}
+
+# Brain input state, consisting of error from target value for position and velocity.
+type ObservableState {
+    # Ball X,Y position
+    ball_x_error: number<-MaxDistancePerStep - RadiusOfPlate .. RadiusOfPlate + MaxDistancePerStep>,
+    ball_y_error: number<-MaxDistancePerStep - RadiusOfPlate .. RadiusOfPlate + MaxDistancePerStep>,
+
+    # Ball X,Y velocity
+    ball_vel_x_error: number<-MaxVelocity .. MaxVelocity>,
+    ball_vel_y_error: number<-MaxVelocity .. MaxVelocity>,
+}
+
+# Action provided as output by policy and sent as
+# input to the simulator
+type SimAction {
+    # Range -1 to 1 is a scaled value that represents
+    # the full plate rotation range supported by the hardware.
+    input_pitch: number<-1 .. 1>, # rotate about x-axis
+    input_roll: number<-1 .. 1>, # rotate about y-axis
+}
+
+# Per-episode configuration that can be sent to the simulator.
+type SimConfig {
+    # Model initial ball conditions
+    initial_x: number<-RadiusOfPlate .. RadiusOfPlate>, # in (m)
+    initial_y: number<-RadiusOfPlate .. RadiusOfPlate>,
+
+    # Model initial ball velocity conditions
+    initial_vel_x: number<-MaxInitialVelocity .. MaxInitialVelocity>, # in (m/s)
+    initial_vel_y: number<-MaxInitialVelocity .. MaxInitialVelocity>,
+
+    # Range -1 to 1 is a scaled value that represents
+    # the full plate rotation range supported by the hardware.
+    initial_pitch: number<-1 .. 1>,
+    initial_roll: number<-1 .. 1>,
+
+    # Target X,Y position
+    target_x: number<-MaxDistancePerStep - RadiusOfPlate .. RadiusOfPlate + MaxDistancePerStep>,
+    target_y: number<-MaxDistancePerStep - RadiusOfPlate .. RadiusOfPlate + MaxDistancePerStep>,
+}
+
+# State transform function definition from absolute to error
+# like commonly found in control theory
+function TransformState (s: SimState): ObservableState {
+    return {
+        ball_x_error: s.target_x - s.ball_x,
+        ball_y_error: s.target_y - s.ball_y,
+        ball_vel_x_error: TargetVelocityX - s.ball_vel_x,
```

Comment:
So renaming the states is good for learning here, but this will affect a few things down the line. This will affect exported brains running in the sim as well as hardware. It's probably fine but what do you think?

Reply:
Good catch — we should add a comment about this. Keeping the same names with different meaning wouldn't help though -- the hardware or sim-with-exported-brain would send ball_x, it would be treated as -ball_x by the brain, and the plate would mysteriously keep doing the opposite of what it should.